src/main/java/com/viaversion/viafabricplus/protocoltranslator/util/JavaPinger.java

```diff
@@ -0,0 +1,123 @@
+
+    private final static Gson GSON = new Gson();
+
+    public static Response ping(InetSocketAddress address, int timeout, int protocolVersion) {
+        try (Socket socket = new Socket(address.getAddress(), address.getPort())) {
+            socket.setTcpNoDelay(true);
+            socket.setSoTimeout(timeout);
+            socket.connect(null, timeout);
+            try (DataOutputStream dataOutputStream = new DataOutputStream(socket.getOutputStream());
+                 DataInputStream dataInputStream = new DataInputStream(socket.getInputStream());
+                 ByteArrayOutputStream b = new ByteArrayOutputStream();
+                 DataOutputStream handshake = new DataOutputStream(b)) {
+                handshake.writeByte(0);
+                writeVarInt(handshake, protocolVersion);
+                writeVarInt(handshake, address.getHostString().length());
+                handshake.writeBytes(address.getHostString());
+                handshake.writeShort(address.getPort());
+                writeVarInt(handshake, 1);
+                writeVarInt(dataOutputStream, b.size());
+                dataOutputStream.write(b.toByteArray());
+                dataOutputStream.writeByte(1);
+                dataOutputStream.writeByte(0);
+
+                int size = readVarInt(dataInputStream);
+                int id = readVarInt(dataInputStream);
+                int length = readVarInt(dataInputStream);
+
+                if (size <= 0 || id != 0 || length <= 0) {
+                    throw new IllegalStateException("No data received from server");
+                }
+                byte[] in = new byte[length];
+                dataInputStream.readFully(in);
+                return GSON.fromJson(new String(in, StandardCharsets.UTF_8), Response.class);
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException("Could not ping " + address + " with protocol " + protocolVersion, e);
+        }
+    }
+
+    private static int readVarInt(DataInputStream in) throws IOException {
+        int i = 0, j = 0;
+        byte k;
+        do {
+            k = in.readByte();
+            i |= (k & 0x7F) << (j++ * 7);
+            if (j > 5) {
+                throw new IOException("VarInt too big");
+            }
+        } while ((k & 0x80) != 0);
+        return i;
+    }
+
+    private static void writeVarInt(DataOutputStream out, int value) throws IOException {
+        while ((value & 0xFFFFFF80) != 0) {
+            out.writeByte((value & 0x7F) | 0x80);
+            value >>>= 7;
+        }
+        out.writeByte(value);
+    }
+
+    /**
+     * Represents the response received from a server after sending a ping request.
+     * The response contains a {@link Version} object providing specific details
+     * about the server's version.
+     */
+    public record Response(Version version) {}
+
+    /**
+     * Represents the version information of a server. This record encapsulates
```

Comment:
No need for such a big javadoc for a very basic record which I believe is not even needed in full

Reply:
IntelliJ AI slop. I'll drop.